docs/map-making/how-to-upgrade-legacy-map-features.md

```diff
@@ -35,70 +19,36 @@ All occurrences of the code before the `>>` must be substituted with the code af
 - `="conditionType" count="XOR"` >> `="conditionType" count="1"`
 - `="conditionType" count="-reset-XOR"` >> `="conditionType" count="-reset-1"`
 - `="isTwoHit" count="true"` >> `="hitPoints" count="2"`
-- `="isTwoHit" value="false"` >> `="hitPoints" value="1"`
-- `="isTwoHit" count="-reset-true"` >> `="hitPoints" count="-reset-2"`
 - `="isTwoHit" count="false"` >> `="hitPoints" count="1"`
+- `="isTwoHit" value="true"` >> `="hitPoints" value="2"`
+- `="isTwoHit" value="false"` >> `="hitPoints" count="1"`
+- `="isTwoHit" count="-reset-true"` >> `="hitPoints" count="-reset-2"`
 - `="isTwoHit" count="-reset-false"` >> `="hitPoints" count="-reset-1"`
 - `="occupiedTerrOf"` >> `="originalOwner"`
 - `="victoryCity" count="true"` >> `="victoryCity" count="1"`
-- `="victoryCity" count="-reset-true"` >> `="victoryCity" count="-reset-1"`
 - `="victoryCity" count="false"` >> `="victoryCity" count="0"`
+- `="victoryCity" value="true"` >> `="victoryCity" count="1"`
+- `="victoryCity" value="false"` >> `="victoryCity" count="0"`
+- `="victoryCity" count="-reset-true"` >> `="victoryCity" count="-reset-1"`
 - `="victoryCity" count="-reset-false"` >> `="victoryCity" count="-reset-0"`
-
-
-### **Procedural alternatives:**
-
-For a more defined change, instead of the aforementioned:
-
-- `attatchment` >> `attachment`
-
-you can change:
-
-- `attatchmentList` >> `attachmentList`
-- `<attatchment` >> `<attachment`
-- `attatchment>` >> `attachment>`
-- `="games.strategy.triplea.attatchments` >> `="games.strategy.triplea.attachments`
-
-- `="techAttatchment` >> `="techAttachment`
-- `="unitAttatchment` >> `="unitAttachment`
-- `="territoryAttatchment` >> `="territoryAttachment`
-- `="canalAttatchment` >> `="canalAttachment`
-- `="rulesAttatchment` >> `="rulesAttachment`
-- `="playerAttatchment` >> `="playerAttachment`
+- `<option name="isParatroop"` >> `<option name="isAirTransportable"`
```

Comment:
Since this is not any longer specifically about upgrading from maps made for TripleA 1.8.0.9, I guess sometime in the past this option was doing something, which may or may not be (exactly) the same that "isAirTransportable" is doing. However, this needs to be verified to be actually the case (we cannot just assume "isAirTransportable" is merely a name change of "isParatroop", because they sound being about the same thing, unless this assumption is documented, which I'm not aware).

Moreover, it may be that there are games in which you have "isParatroop" set to "false" in a unit attachment after "isAirTransportable" set to true (garbage code, which I can tell is generally frequent in TripleA games). In this case, if in the TripleA for which the game was made the "isParatroop" was ineffective, you are substantially bugging the game, changing the behaviour of "isAirTransportable" from "true" to "false" (unburying bad coding).

In theory, to be exact, here you would need to find out the range of TripleA versions for which "isParatroop" was doing nothing, and telling that upgrading from those need to have the option deleted, to be sure to keep the behaviour unchanged (the upgrader would need to check the minimum version of the game, and delete or otherwise, depending on it).

Reply:
@Cernelius I don't think we can assume that if a map is at version 1.8 that it was correctly upgraded to 1.8. Specifically, if we see 'isParatroop', does that mean the new property was simply added, or does it mean it was never added?

IMO it's better to upgrade and allow the map maker to spot and remove duplicate options if then present. That handles the potential 'true' / 'false' case which would already have been a potential problem.

Here is the diff of when 'isAirTransportable' was added: 

commit d56265ad9bef86651ccbcd310f3964b78188d5b0
Author: comradekev <comradekev@yahoo.com>
Date:   Thu Jun 3 19:36:19 2010 +0000

    * fix bug 2998846 kamikaze aircraft not working * add isAirTransport(able) to support air dropping units * fix bug 2994004 air landing on newly produced carrier in multi-national SZ disappears * feature request (2799840) add tech tokens to status tab

    git-svn-id: https://svn.code.sf.net/p/triplea/code/trunk/triplea@2785 a4c711aa-641f-0410-879d-9b464e52b682

Reply:
I guess there's no need to worry too much about these things, since this is a mapmaker aid, but certainly I believe that, if we were ourselves making mass-changes, mass changing all "isParatroop" to "isAirTransportable", after many years in which the original has been ineffective, would be very risky.

It may be interesting to see a summary of all instances of "isParatroop" that have been removed in the repository, back then when the mass-deletion was made (I'm assuming it was).

I'll still point out that we should not assume that "isParatroop" used to do the same that "isAirTransportable" is doing just because the name seems to be about the same thing. We shouldn't really infer anything from the name itself, as that is merely a label. So such a change really needs confirmation that "isAirTransportable" is indeed merely a renamed version of "isParatroop". If it is not, then we would be changing, not upgrading, also games made when the "isParatroop" option was non deprecated.

All this based on the assumption that what I know and documented is right, since I've not tested any of this, at least recently.

Reply:
Given the time of when 'isAirTransportable' was introduced, I'm pretty sure it was not mass replaced. It was left to individual map makers to make the update and instead their maps were broken by leaving in a useless property that while the game still ran, the 'isParatroop' no longer did anything. So map makers would have updated it or their maps were changed.


>  So such a change really needs confirmation that "isAirTransportable" is indeed merely a renamed version of "isParatroop". If it is not, then we would be changing, not upgrading, also games made when the "isParatroop" option was non-deprecated.

Check the referenced commit, it replaces isParatroop. I would guess if we see 'isParatroop', it's more likely that it was never replaced rather than was replaced but left in. I find that a far more likely scenario. Beyond that, if isParatroop is in the map, then 'isAirTransportable' was the desired behavior. A map not having 'isAirTransportable' would then not be functioning correctly. I'd reject the statement that if a map made it to 1.8 and someone continued to overlook that, then 'oh well', the map was never supposed to have 'isParatroop' to begin with.

Reply:
> I guess there's no need to worry too much about these things, since this is a mapmaker aid, but certainly I believe that, if we were ourselves making mass-changes, mass changing all "isParatroop" to "isAirTransportable", after many years in which the original has been ineffective, would be very risky.

I think the days of mass changes are behind us. The migration is breaking to old engines and the benefit is non-existent now that 2.3 and beyond will be able to handle legacy values.